readme.md
### Requirements
- running server of `rabbitmq`
- running server of `Lumen`
- 'Nginx' proxy port for listening 
---
### Running services
```
bash
sudo vim /lib/systemd/system/havijMQ.service
sudo systemctl daemon-reload
sudo systemctl enable havijMQ.service
sudo systemctl start havijMQ.service
sudo systemctl restart havijMQ.service
sudo systemctl status havijMQ.service
sudo journalctl -f --no-pager -n 1000 --output=cat -u havijMQ
```

CHANGELOG.md
# Change Log

#### [0.0.0] - 2021-01-10
#### Added
- test
#### Changed
- test
#### Fixed
- test

producer.py
import json
import os

import pika
from aiohttp import web
from dotenv import load_dotenv


async def handle(request):
    data = await request.post()
    if 'Authorization' not in request.headers:
        return web.HTTPUnauthorized()
    response = dict(data)
    try:
        jsons = await request.json()
        response.update(jsons)
    except:
        pass

    response.update({
        'method': request.method,
        'url': request.path,
        'token': request.headers['Authorization'],
    })
    channel.basic_publish(exchange='', routing_key=CONFIG['RabbitMQQueue'], body=json.dumps(response))
    return web.HTTPOk()


async def run():
    global channel, CONFIG
    load_dotenv()
    CONFIG = {
        'RabbitMQServer': os.getenv("RabbitMQServer", 'localhost'),
        'RabbitMQPort': os.getenv("RabbitMQPort", '5672'),
        'RabbitMQQueue': os.getenv("RabbitMQQueue", 'test'),

        'NginxProxyPort': int(os.getenv("NginxProxyPort", '8080')),
    }

    connection = pika.BlockingConnection(pika.ConnectionParameters(host=CONFIG['RabbitMQServer'], heartbeat=0))
    channel = connection.channel()
    channel.queue_declare(queue=CONFIG['RabbitMQQueue'])

    app = web.Application()
    app.router.add_route('GET', '/{tail:.*}', handle)
    app.router.add_route('POST', '/{tail:.*}', handle)
    app.router.add_route('PUT', '/{tail:.*}', handle)
    app.router.add_route('DELETE', '/{tail:.*}', handle)

    return app


fake_load.py
import asyncio

import aiohttp
import async_timeout
import tqdm


async def fetch(session, url):
    with async_timeout.timeout(200, loop=session.loop):
        async with session.get(url) as response:
            return await response.text()


async def run(r):
    url = "http://192.168.5.34/exam"
    # The default connection is only 20 - you want to stress...
    conn = aiohttp.TCPConnector(limit=200)
    tasks, responses = [], []
    async with aiohttp.ClientSession(connector=conn) as session:
        tasks = [asyncio.ensure_future(fetch(session, url)) for _ in range(r)]
        # This will show you some progress bar on the responses
        for f in tqdm.tqdm(asyncio.as_completed(tasks), total=len(tasks)):
            responses.append(await f)
    return responses


number = 100000
loop = asyncio.get_event_loop()
loop.run_until_complete(run(number))


fake_generator.py
import json

import pika
from tqdm import tqdm

message = {
    'token': 'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJhdWQiOiIxIiwianRpIjoiNjU3M2I5ZDA0YjljNDA0NjQ4ZjUwMGJkZjYzZTNlNTU3NTk0ZjU4OWRjN2JkZTcwNWE1MjYwNDQ0YmI3MTg1MmRkZWNkODEzNTE2NWMyNGIiLCJpYXQiOjE2MTAyNzE4ODYsIm5iZiI6MTYxMDI3MTg4NiwiZXhwIjoxNjQxODA3ODg2LCJzdWIiOiIzODkyNzAiLCJzY29wZXMiOltdfQ.NxudRynsXYSwWRTxUsUXJbXgB5cLC0LJYMWIzuaq0BJsAXpp4yOpJ7e0tfkNwS8q-UcFPq5BW3CnffpPLsF8wWhveUqBsYI_Ex5mGl5tc65HojhDdt2_kCewo30GRXmq0JNfewV-OG2xC1oBaZU1m2JUBHLxTxI6rltZ0WnK6y5mHjdU7isXiZqOStrftIgWKbBxFWzmPqq8-ZlRIkpsAYuseZdybh0w1XCKq-EywbP9j3ARovz8w4seuAj8OKsfUEShueA4Lhur8grfPfiDfzmA-56xM3F8AuOrfpRVU0FRMHQ93JchYqdLwS7aA9mdT2D13ixr-IgrcYoeJtzYAXZKVKD_loaGr0HHi5DRJDQfs8kdu8dgtrKpFFO4GWEkzd1qY_zA2RcO2cLB_lFPjXY7E8JNwxGHyrkKKbXJcaaKMKVT2CiHb-wZUPnos-f74X7l2dbXzHThNvvZ0facfJawpSfbAdJn-0X5I1nmCGO6Q0xjxQLMBy7j0lf3wJ8UZi-l1eDDyM_6cWrHMhVH8lOMc1mj8eMTF9y0k2o7mJc7pZW49gPo8Eh2ky5uWgIAEH-kZgN8voXksExq3kBqH1IlSB6l7O-VxQ7jN8jzRJG2xS6HPPFPo6FDQYsTcTfrw_RgheONfRaJ7T8geJ_hit7JugH92veJAUaMRHOKtPQ',
    'url': '/exam?id=5ffabb4e0d7c702a56046835',
    'method': 'get',
    'jigil': 'fingil',
    'nini': 'mimi',
}
queue_name = 'test'
connection = pika.BlockingConnection(pika.ConnectionParameters(host='localhost'))
channel = connection.channel()
channel.queue_declare(queue=queue_name)

reqs = 500
bar = tqdm(total=reqs, desc='making requests ', unit=' request', ncols=250)
for i in range(reqs):
    channel.basic_publish(exchange='', routing_key=queue_name, body=json.dumps(message))
    bar.update(1)
    # print(i)
bar.close()
connection.close()


consumer.py
import datetime
import json
import logging
import os
import sys
import threading
from datetime import date
from threading import Thread

import pika
import requests
from dotenv import load_dotenv


def main():
    config()
    balance()


def config():
    global CONFIG, HEADERS
    load_dotenv()
    log_levels = ['--DEBUG', '--INFO', '--WARNING', '--ERROR', '--CRITICAL']
    USAGE = '''
RabbitMQ consumer APP

--help          for help
--[level]       for log level {levels}'''.format(length='multi-line', levels=log_levels)

    HEADERS = {
        'Accept': 'application/json',
        'Cookie': 'nocache=1',
        'Accept-Encoding': 'utf-8'
    }

    opts = [opt for opt in sys.argv[1:] if opt.startswith("--")]
    if "--help" in opts:
        print(USAGE)
        exit(0)
    try:
        log_level = (str(list(set(log_levels) & set(opts))[0])).replace('--', '')
    except IndexError:
        log_level = (str(log_levels[-1])).replace('--', '')

    today = date.today()
    log_file = str(today) + '.log'
    log_format = '%(levelname)s %(asctime)s - %(message)s'
    logging.basicConfig(filename=log_file,
                        format=log_format,
                        level=getattr(logging, log_level))

    CONFIG = {
        'SIMULTANEOUS_THREADS': int(os.getenv("SIMULTANEOUS_THREADS", 20)),

        'RabbitMQServer': os.getenv("RabbitMQServer", 'localhost'),
        'RabbitMQPort': os.getenv("RabbitMQPort", '5672'),
        'RabbitMQQOS': int(os.getenv("RabbitMQQOS", 10)),
        'RabbitMQQueue': os.getenv("RabbitMQQueue", 'test'),

        'LumenSchema': os.getenv("LumenSchema", 'http'),
        'LumenServer': os.getenv("LumenServer", 'localhost'),
        'LumenPort': int(os.getenv("LumenPort", 80)),
    }
    print(
        "======== Running on {} ========\n"
        "(Press CTRL+BREAK to quit)".format(CONFIG['LumenPort'])
    )


def balance():
    threads = []
    try:
        while True:
            while threading.activeCount() > CONFIG['SIMULTANEOUS_THREADS']:
                pass
            threads = [t for t in threads if t.is_alive()]
            threads.append(Thread(target=connection))
            threads[-1].start()

    except KeyboardInterrupt:
        exit()


def connection():
    try:
        connection = pika.BlockingConnection(pika.ConnectionParameters(host=CONFIG['RabbitMQServer'], heartbeat=60))
        channel = connection.channel()
        channel.queue_declare(queue=CONFIG['RabbitMQQueue'])
        channel.basic_qos(prefetch_count=CONFIG['RabbitMQQOS'])
        channel.basic_consume(queue=CONFIG['RabbitMQQueue'], on_message_callback=send_request, auto_ack=True)
        try:
            logging.info('thread started listening consuming')
            channel.start_consuming()
        except:
            logging.critical('thread consumption fialed')
            channel.stop_consuming()
    except:
        logging.critical('connection to MQ failed')

    connection.close()


def send_request(ch, method, properties, body):
    start = datetime.datetime.now()
    message = json.loads(body)
    HEADERS['Authorization'] = message['token']
    try:
        response = requests.request(message['method'], CONFIG['LumenSchema'] + CONFIG['LumenServer'] + message['url'],
                                    headers=HEADERS, data=message)
        status_code = response.status_code
        if status_code != 200:
            logging.critical('request non 200')
        data = json.loads(response.content)
        logging.debug(status_code)
        logging.debug(data)
    except:
        logging.critical('request failed')
        connection = pika.BlockingConnection(pika.ConnectionParameters(host=CONFIG['RabbitMQServer']))
        channel = connection.channel()
        channel.basic_publish(exchange='', routing_key=CONFIG['RabbitMQQueue'], body=json.dumps(message))
        connection.close()

    end = datetime.datetime.now()
    elapsed = end - start
    logging.debug('time spend on a single request: ' + str(elapsed.microseconds))


main()
